### DISCUSSION TITLE
How to use aws-wrapper jdbc driver in Quarkus based Keycloak?

### DISCUSSION THREAD

**Discussion by fredxing** [keycloak/keycloak#15292] [Archive]
We need to run Keycloak (Quarkus based) in AWS EKS and want to use IAM authentication to access RDS-Postgresql. AWS has released a JDBC driver to handle that (https://github.com/awslabs/aws-advanced-jdbc-wrapper). However we couldn't make Keycloak to use this driver. We have tried to config KC_DB_URL to the url (jdbc:was-wrapper:postgresql://...) because the url protocol is not starting with jdbc:postgresql, so the url is not recognized and fall back to default one. Does anybody know how to do that? 

ENV KC_DB=postgres
ENV KC_DB_URL=jdbc:was-wrapper:postgresql://myserver:5432/mydb

**Comment by sschu** [comment]
Keycloak takes its database driver from Quarkus. You would probably have to create a Quarkus extension around this driver for this to work. However, I never did this myself so I cannot really help you. You could look into Quarkus and see how the available jdbc driver extensions look like...

**Reply by fredxing** [reply]
Thanks for the suggestion. I will give it a try.

**Comment by fredxing** [comment]
I found if I specify the db-driver to the aws driver at build time (with db=postgres), I'm able to see it was picked up at runtime. However, postgresql driver was not registered anymore. Since the aws driver is a wrapper of postgresql driver, so I got 'No suitable driver' for jdbc:postgresql://. I also tried to add postgresql driver as a provider but still couldn't find it. It seemed to me the DriverManager was loaded in different class loader that had only one driver at a time. Does anyone have any suggestions on this?

**Comment by fredxing** [comment]
I finally made it working by changing the aws wrapper driver to reload (register) the postgresql jdbc driver after register itself. Here is the Dockerfile I used,

```
###
# Keycloak (Quarkus based) optimized docker image build file
#    
###
FROM quay.io/keycloak/keycloak:20.0.0 as builder

# build options
ENV KC_METRICS_ENABLED=true
ENV KC_HEALTH_ENABLED=true
ENV KC_HTTP_RELATIVE_PATH=/auth
ENV KC_DB=postgres
ENV KC_DB_DRIVER=software.amazon.jdbc.Driver
ENV KC_TRANSACTION_XA_ENABLED=false
ENV KC_CACHE_CONFIG_FILE=cache-ispn-jdbc-ping.xml

# copy the custom cache config file into the keycloak conf dir
COPY ./cache-ispn-jdbc-ping.xml /opt/keycloak/conf/
# add aws jdbc driver and dependencies (with the fixed version of aws driver)
COPY ./libs/*.jar /opt/keycloak/providers/
RUN curl -sL https://s3.amazonaws.com/rds-downloads/rds-ca-2019-root.pem -o /opt/keycloak/rds-ca-2019-root.pem
# run optimizing build
RUN /opt/keycloak/bin/kc.sh build

FROM quay.io/keycloak/keycloak:20.0.0
COPY --from=builder /opt/keycloak/ /opt/keycloak/
WORKDIR /opt/keycloak

ENTRYPOINT ["/bin/bash", "-c", "printenv | sort && /opt/keycloak/bin/kc.sh --verbose start --optimized --log-level=INFO"]
```

**Reply by ppietruha** [reply]
@fredxing I used Yours dockerfile - with no luck. Did You do some extra steps? I see keycloak registered just aws-wrapper driver.

**Reply by fredxing** [reply]
I made a change in AWS wrapper Driver class to reload the postgresql driver. Hope this can help you (see below).

public static void register() throws SQLException {
   ......
    //check postgresql driver
    try {
      DriverManager.getDriver("jdbc:postgresql:");
    } catch (SQLException exp) {
      try {
        Class.forName("org.postgresql.Driver");
        DriverManager.getDriver("jdbc:postgresql:");
        LOGGER.info("Loaded Postgresql JDBC driver.");
      } catch (ClassNotFoundException expp) {
        throw new IllegalStateException("Postgresql Driver not found.");
      }
    }
........

**Comment by ahus1** [comment]
For Keycloak 24, it is now part of the official docs how to configure Amazon JDBC wrapper: https://www.keycloak.org/server/db#preparing-keycloak-for-amazon-aurora-postgresql

**Reply by codespearhead** [reply]
@sschu Since you're participating in this discussion, could you mark the above comment as the accepted answer? This will prevent it from showing up as unanswered in the GitHub discussions tab.

I've also included the git blame references [2][3] to support this, as the information may have been overlooked in the Keycloak 24 release notes.

***

Related: [4][5].

***

[1] https://docs.github.com/en/discussions/managing-discussions-for-your-community/moderating-discussions#marking-a-comment-as-an-answer
[2] https://github.com/keycloak/keycloak/blame/9404b888d1f85184d95110d7019f0b5632222822/pom.xml#L163
[3] https://github.com/keycloak/keycloak/commit/7135b4ec4c582e3b0405122a779335a6a847c80a
[4] https://github.com/keycloak/keycloak/discussions/33995
[5] https://github.com/keycloak/keycloak/discussions/33995#discussioncomment-10983043

**Reply by sschu** [reply]
I cannot mark it as accepted answer, probably because I did not ask the question in the first place. However, I can close the discussion if that is fine for you.

**Reply by codespearhead** [reply]
Thanks! I think closing the discussion is good enough then.
